### PR TITLE
[DOC] [PDF] Avoid pi digits overflowing into right margin

### PR DESCRIPTION
**Pull request by jfbu**
#### References to other Issues or PRs

N/A

#### Brief description of what is fixed or changed

Extra latex via raw directive to avoid this

Capture d’écran 2024-07-27 à 13 58 20

It will show like this: (scale is not same due to github resizing above png to fit)

Capture d’écran 2024-07-27 à 14 23 40

#### Other comments
None

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by jfbu** [comment]
ah wait, do I read correctly
```
SPHINXVER    = 0.5
```
in doc/Makefile? It can't be that you are using Sphinx `0.5`... (proof of this is PDF Doc at 1.13.1 anyhow).

Besides this ``SPHINXVER`` seems to use din Makefile.

File: doc/requirements.txt [hunk 1]
```diff
@@ -5,7 +5,7 @@ matplotlib-inline
 mpmath
 myst-parser
 scipy
-Sphinx
+Sphinx>=3.5.0
```
Comment (asmeurer) [inline doc/requirements.txt:8]: I'm sure we require a much newer version of Sphinx than this.

**Comment by asmeurer** [comment]
The Sphinx version in the makefile is probably the version of Sphinx that was used back when the makefile was first generated. The docs require a relatively new version of Sphinx, and in practice you should always use the latest version.

File: doc/src/tutorials/intro-tutorial/basic_operations.rst [hunk 2]
```diff
@@ -129,13 +129,22 @@ To evaluate a numerical expression into a floating point number, use
     >>> expr.evalf()
     2.82842712474619
 
+.. raw:: latex
+
+   \begingroup
+   \sphinxsetup{verbatimforcewraps}
```
Comment (asmeurer) [inline doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: Should we just add this to the latex configuration? I'm sure this isn't the only example that overflows in the PDF.

Comment (jfbu) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: It could be added to global configuration like this:
```
latex_elements = {
    ...other things...,
    'sphinxsetup': 'verbatimforcewraps',
}
```

Normally long code lines are wrapped without this setting satisfactorily because break points are found at least at spaces and at various non-letter punctuation.  The setting is needed for things such as this case of a long string of digits, one can also imagine DNA sequence of such types of things.  It causes Sphinx to use more complex LaTeX macro layer which remaps all mark-up applied by Pygments styling to apply character per character, inserting then breakpoints everywhere.

If you can get a hold of the LaTeX buid log file, checking for "overfull \hbox" should help locate other cases and see how frequent or numerous they are  I tried perusing the PDF but did not immediately find other instances, now, I am only a human not an AI... (which would immediately locate in many thousands pages the strings overflowing into margins...).

File: doc/requirements.txt (hunk 1 above)
Comment (jfbu) [reply doc/requirements.txt:8]: I could see your build uses the quasi latest version as the PDF for 1.13.1 displayed some styling improvements in PDF output for admonitions, which were very recently merged (at Sphinx 7.4.0, which is very recent -- the latest release was done today and is 8.0.0 which has dropped deprecated branches).

Comment (jfbu) [reply doc/requirements.txt:8]: I added this `>=3.5.0` only to make sure `verbatimforcewraps` existed but if the advice to contributors/testers for builds is always to use latest Sphinx, then this should be dropped of course.

File: doc/src/tutorials/intro-tutorial/basic_operations.rst (hunk 2 above)
Comment (asmeurer) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: There is a build log available if you build the pdf (see https://docs.sympy.org/latest/contributing/new-contributors-guide/build-docs.html#pdf-documentation). Or you could modify the GitHub actions to put the log as an artifact. I seem to remember that there are quite a few overfull \hbox errors in the log the last time I looked at it.

Comment (jfbu) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: Ok, I managed to build locally after installing a Chrome Legacy for my old macos.  The svg to pdf conversions failed despite chromium being in expected place but a least this did not cause a fatal error as with my Inkscape 0.92.2 and I could then try out the xelatex builds.  Of course it would be probably annoying to add pdf's to the git, but if these images are not dynamic, then perhaps you may consider using glob pattern in rst sources (not sure how/if it works with myST) so that Sphinx knows to use the pre-prepared pdf's for LaTeX. (but sometimes pdf inclusion has [problems](https://tex.stackexchange.com/questions/106964/could-not-insert-pdf-graphics) with xelatex).

The Makefile should add `logo` as dependency to `latex` targets, because as I did not build the html docs, the `_build/logo` was missing, and the produced LaTeX file expects to find some `png`'s in-there and build fails without them.

Turns out I did the build using my branch for PR #26866, so the end of log was already missing probably dozens (or dozens of dozens) `Overfull \hbox` generated from general Index, but I did find some interesting ones, but only 3 of the type of the pi digits overflow. (I discuss this in more detail next).  Before this, I had checked Sphinx and now I remember that the extra overhead added by `verbatimforcewraps` is not excessive: it adds an extra "measuring phase" to all renderings of codelines, which may not be that time costly anyhow (I did not time this because the build takes time and I had to address other issues first such as the logo business), and if that phase detects an overflow (which for Sympy is very rare) then it does the more hacky things I alluded too (the whole business being already quite deep core TeX...).

In my build (which as i said is missing some pdf's, so page numbering maybe off) the few instances similar to pi digits were on pages 1380, 1838, and 2285 (and more of same type a bit after I think).  Here are screenshots

page 1380
Capture d’écran 2024-07-30 à 11 11 32

page 1838
Capture d’écran 2024-07-30 à 11 12 44

page 2235
Capture d’écran 2024-07-30 à 11 25 52

The last case is irreparable.  The nature of the thing (Maple like output) can only be wrapped by human intervention, impossible to do this automatically at Sphinx level.  The first two would be solved by local or global `verbatimforcewraps`.

When checking `Overfull \hbox` the line number in LaTeX file is indicated which allows to locate source too.  The page number in PDF is most of the time the next bracketed one, but sometimes it is actually on next page, for example when a tabular is too long to fit on page and is moved to next one during latex "output routine".  The log line indicated the amount of overfull and you only need to worry (for letter sized paper) when it exceeds 70pt.  Above that you do lose characters beyond page limits.

Some overfull are located in a few tables, and probably could be fixed by suitable usage of Sphinx  [``tabularcolumns`` directive](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-tabularcolumns).  Only two or three tables are involved the worst one is this but I have lost on which page, the other tables had more benign problems.
Capture d’écran 2024-07-30 à 11 17 42

Then they are overfull caused by too long displayed equations. This can only be fixed in sources (I don't know of a robust LaTeX solution to automatically break lines in math displays, a package breqn I think exist but causes problems in some circumstances, perhaps worth a try, not tested).

Here is an example (5.8. Topics page 837)
Capture d’écran 2024-07-30 à 11 03 37

There are a few other math display problems not very numerous.

Then there is sometimes a problem with an hyperlink which overflows in margin, normally hyperref knows how to break it but it may not find suitable breakpoints (and I think never break at slashes), but this is important only for print version, as numeric PDF will have the correct target on hover with mouse.

The last class of problems may need some ticket opening at https//github.com/sphinx-doc/sphinx.
I located a dozen cases like this one (one of the worst):

Capture d’écran 2024-07-30 à 11 22 05

A few more:
Capture d’écran 2024-07-30 à 11 10 45

Capture d’écran 2024-07-30 à 11 07 30

Capture d’écran 2024-07-30 à 11 20 29

Such problemes deserve if you have time opening a ticket at Sphinx reusing one of abvoe images.

Finally, some Missing characters are reported, see #26877.

Comment (jfbu) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: I intend to become much more familiar with Sympy in future and pretty printed output looks like a pretty cool feature...

Comment (jfbu) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: Tomorrow I will have a look at the few tables where I identified too narrow columns.  Suitable usage of `tabularcolumns` Sphinx directive will solve that.

Comment (jfbu) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst:135]: > Should we just add this to the latex configuration? I'm sure this isn't the only example that overflows in the PDF.

Although I found only two further examples with overflowing digits I agree it is a bit painful to have to maintain raw latex directive in sources.  So I am going to revert what I did for the initial spotted case, the one of pi digits, and add the `verbatimforcewraps` in conf.py.  Tomorrow I will test if I see any significant increase in build time.

Comment (jfbu) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst]: (I don't know if Sympy process is via squash/rebase during merge to master, but I am expecting it so I do not worry on my accumulating small commits).  I plan to push to this branch some more commit improving tabulars in PDF.  And I will raise at some point an upstream Sphinx issue about documenting objects with no room left on right side for parameters in the signature. It seems we should in those cases shift to under the long object name, simply, regaining ample room for arguments (especially if they come with type annotations this is quite needed).

Comment (oscarbenjamin) [reply doc/src/tutorials/intro-tutorial/basic_operations.rst]: > don't know if Sympy process is via squash/rebase during merge to master, but I am expecting it so I do not worry on my accumulating small commits

See https://github.com/sympy/sympy/issues/18548

**Comment by moorepants** [comment]
@jfbu, this is a side question (sorry for hijacking the PR thread), but do you know of a way to handle the issue of latex output from SymPy code having an overflow:
image

Is this something Sphinx can deal with or do we need to control it with the SymPy printers?

**Comment by jfbu** [comment]
@moorepants If your input is as I find in Sympy doc/ with ``math`` Sphinx directive, then a fix by manual intervention is possible via adding one `\\` inside the math content.  For example, 
```diff
-28. .. math:: p < q -1 \wedge t = 0 \wedge \phi = 0 \wedge 0 < s \wedge 0 < b^{*} \wedge 0 \leq c^{*} \wedge \pi c^{*} < \lvert{\operatorname{arg}\left(\omega\right)}\rvert \wedge \lvert{\operatorname{arg}\left(\omega\right)}\rvert < \pi \left(m + n - p + 1\right) \wedge C_{1} \wedge C_{2} \wedge C_{10} \wedge C_{14} \wedge C_{15}
+28. .. math:: p < q -1 \wedge t = 0 \wedge \phi = 0 \wedge 0 < s \wedge 0 < b^{*} \wedge 0 \leq c^{*} \wedge \pi c^{*} < \lvert{\operatorname{arg}\left(\omega\right)}\rvert \wedge \lvert{\operatorname{arg}\left(\omega\right)}\rvert < \\ \pi \left(m + n - p + 1\right) \wedge C_{1} \wedge C_{2} \wedge C_{10} \wedge C_{14} \wedge C_{15}
```
as I am currently trying out (and it works) as a patch to doc/src/modules/integrals/g-functions.rst, see next image and compare to the one if my previous comment.

Capture d’écran 2024-08-01 à 10 19 39

I am checking the HTML build and the output looks ok (to me) there (contrast the 28 and 29 which I edited and the 26 and 27 which HTML renders with a scroll-bar; they needed no intervention in my patch as they fit in text area in a letter wide page):

Capture d’écran 2024-08-01 à 10 35 03

(@asmeurer I will push at some point  a commit to this branch with my manually added `\\`'s, with the understanding it will be reverted if changes to HTML output as illustrated above are not acceptable.)

- There is LaTeX package `breqn` which promises "Overlong equations will be broken automatically".  Unfortunately this seems to mean using specific environements, i.e. to modify the way Sphinx renders the ``math`` directive.  As the package also breaks commonly found math mark-up, I have not invested much time yet researching its usability. 
- An automatic insertion by Sphinx of `\\` (the latter works because Sphinx uses `split` environment inside `equation*` environment) looks to me a bit challenging because due to control sequences it is hard to tell when line will be long, and where to put a `\\`.  Would be a good challenge for a useful Sphinx extension.  And it could impact solely the LaTeX.
- Doing this purely on LaTeX side by Sphinx itself would mean writing a LaTeX package solving a problem which apparently has no pre-existing solution in LaTeX world (but I may be not informed enough).  Of course we do that routinely at Sphinx, but this one appears hard :smiley:.

**Comment by moorepants** [comment]
> If your input is as I find in Sympy doc/ with math Sphinx directive, then a fix by manual intervention is possible via adding one \\ inside the math content. For example,

No, my example is not from the SymPy docs and isn't using the `..math::` directive (at least not directly). It uses jupyter-sphinx which somehow captures the latex representation of a sympy code output.

**Comment by jfbu** [comment]
> > If your input is as I find in Sympy doc/ with math Sphinx directive, then a fix by manual intervention is possible via adding one \ inside the math content. For example,
> 
> No, my example is not from the SymPy docs and isn't using the `..math::` directive (at least not directly). It uses jupyter-sphinx which somehow captures the latex representation of a sympy code output.

I was expecting something like that from hearing about "Sympy pretty-printing".  Sadly, I am yet quite ignorant of this eco-system.  Working on it, but not yet reading the documentation...

**Comment by jfbu** [comment]
> No, my example is not from the SymPy docs and isn't using the `..math::` directive (at least not directly). It uses jupyter-sphinx which somehow captures the latex representation of a sympy code output.

Perhaps https://github.com/sympy/sympy/blob/d32e886fe1cd031adad6f5c24a0bc91e33e35bbf/sympy/simplify/hyperexpand_doc.py#L13-L16

is an example of a somewhat similar instance here.  I identified a problem in the `.tex` file which I wanted to solve by addition of a `\\` but the auto-generation from source via the `sympy.printing.latex.latex()` method prevents any such manual intervention.  It is at the level of this method that the insertion of the `\\` should happen.

**Comment by moorepants** [comment]
We could add a flag to the `latex()` printer that inserts line breaks and then this could be turned on with `init_printing()`. I was just wondering if Sphinx new how to manage such a thing automatically already.

**Comment by jfbu** [comment]
I will later take care of some tabular's and this will be it for this PR.

The overwhelming majority of remaining overflowing originates in two contexts:
- the long signatures.  I opened https://github.com/sphinx-doc/sphinx/issues/12723
- the general index. This is fixed by https://github.com/sympy/sympy/pull/26866

Regarding the effect of `verbatimforcewraps`:

Capture d’écran 2024-08-01 à 11 31 12

and

Capture d’écran 2024-08-01 à 11 38 24

but pretty-printing fractions produces things like this:
Capture d’écran 2024-08-01 à 11 41 06

Already prior to this PR long lines were wrapped individually at spaces anyhow (compare same example in [earlier comment](https://github.com/sympy/sympy/pull/26867#discussion_r1696735523)), so only the long `----...` were not.  So this does not solve it but in my view does not make terribly worse either.

**Comment by jfbu** [comment]
> We could add a flag to the `latex()` printer that inserts line breaks and then this could be turned on with `init_printing()`. I was just wondering if Sphinx new how to manage such a thing automatically already.

Only to confirm that no, Sphinx does not know how to insert line breaks automatically apart from outputting `\begin{split}...\end{split}` (for a single equation rendered via ``.. math::``).

**Comment by jfbu** [comment]
I pushed a concluding commit to allow wrapping at underscores in long attribute names which end up in the 3cm wide first columns of the Attributes tables.  Probably one will like to increase anyhow 3cm to 3.5cm (I will add a suggested change to apply if this is preferred) but here is how the test table renders with the line wrap now allowed:

Capture d’écran 2024-08-01 à 19 42 57
 
We see here 3cm is unfortunate for this table because `dyn_implicit_mat` and `dyn_implicit_rhs` behave differently due to wider `m` letter despite same letter count.  Probably at `3.1cm` (untested) there is no wrapping.  Also there would be no such difference if the attributes were rendered in monospace font, i.e. were marked as inline literals or transformed into such by the `ext/docscrape_sphinx.py` code.  The Sphinx mark-up activates wrapping for inline literals so the conf.py latex preamble setting would not be needed.

I am planning no more activity on this PR.

File: doc/ext/docscrape_sphinx.py [hunk 3]
```diff
@@ -129,6 +129,8 @@ def _str_member_list(self, name):
             #     out += [''] + autosum
 
             if others:
+                out += [r'.. tabularcolumns:: p{3cm}p{\dimexpr\linewidth-3cm-4\tabcolsep\relax}']
```
Comment (jfbu) [inline doc/ext/docscrape_sphinx.py:132]: perhaps better (see [comment](https://github.com/sympy/sympy/pull/26867#issuecomment-2263662908)):
```suggestion
                out += [r'.. tabularcolumns:: p{3.5cm}p{\dimexpr\linewidth-3.5cm-4\tabcolsep\relax}']
```

File: doc/ext/docscrape_sphinx.py [hunk 4]
```diff
@@ -129,6 +129,8 @@ def _str_member_list(self, name):
...
+                out += ['.. rst-class:: longtable']
```
Comment (jfbu) [inline doc/ext/docscrape_sphinx.py:133]: The tabularcolumns directive has impact only on latex builder. The added longtable class in HTML output does not harm.

File: doc/src/conf.py [hunk 5]
```diff
@@ -372,7 +372,13 @@
 \usepackage{letltxmacro}
 \LetLtxMacro\OldLaTeX\LaTeX
 \AtBeginDocument{\DeclareRobustCommand{\LaTeX}{\text{\OldLaTeX}}}
-'''
+\let\OldUnderscore\_
+\makeatletter
+\AtBeginDocument{\sbox\sphinxcontinuationbox{\spx@opt@verbatimcontinued}}
+\makeatother
+\protected\def\_{\OldUnderscore\discretionary{}{\sphinxafterbreak}{}}
```
Comment (jfbu) [inline doc/src/conf.py:379]: This mantra is to allow TeX to break after an underscore (it does work also if `\_` is encountered in math mode).

File: doc/src/modules/integrals/g-functions.rst [hunk 6]
```diff
@@ -365,18 +365,18 @@ as follows:
...
+33. .. math:: m = 0 \wedge \phi = 0 \wedge u = v + 1 \wedge 0 < n \wedge 0 < c^{*} \wedge 0 \leq b^{*} \wedge \pi b^{*} < \lvert{\operatorname{arg}\left(\sigma\right)}\rvert \wedge \lvert{\operatorname{arg}\left(\sigma\right)}\rvert < \\ \pi \left(b^{*} + 1\right) \wedge C_{1} \wedge C_{3} \wedge C_{12} \wedge C_{14} \wedge C_{15}
+34. .. math:: n = 0 \wedge \phi = 0 \wedge u < v -1 \wedge 0 < m \wedge 0 < c^{*} \wedge 0 \leq b^{*} \wedge \pi b^{*} < \lvert{\operatorname{arg}\left(\sigma\right)}\rvert \wedge \lvert{\operatorname{arg}\left(\sigma\right)}\rvert < \\ \pi \left(s + t - u + 1\right) \wedge C_{1} \wedge C_{2} \wedge C_{12} \wedge C_{14} \wedge C_{15}
+35. .. math:: m = 0 \wedge \phi = 0 \wedge v + 1 < u \wedge 0 < n \wedge 0 < c^{*} \wedge 0 \leq b^{*} \wedge \pi b^{*} < \lvert{\operatorname{arg}\left(\sigma\right)}\rvert \wedge \lvert{\operatorname{arg}\left(\sigma\right)}\rvert < \\ \pi \left(s + t - v + 1 \right) \wedge C_{1} \wedge C_{3} \wedge C_{12} \wedge C_{14} \wedge C_{15}
```
Comment (jfbu) [inline doc/src/modules/integrals/g-functions.rst:379]: in all of the above I simply added one `\\` right before `\pi` in last portion of each equation.

File: sympy/physics/wigner.py [hunk 7]
```diff
@@ -541,10 +541,12 @@ def wigner_6j(j_1, j_2, j_3, j_4, j_5, j_6, prec=None):
...
 
+         \begin{aligned}
          \operatorname{Wigner6j}(j_1,j_2,j_3,j_4,j_5,j_6)
-          =\operatorname{Wigner6j}(j_1,j_5,j_6,j_4,j_2,j_3)
-          =\operatorname{Wigner6j}(j_4,j_2,j_6,j_1,j_5,j_3)
-          =\operatorname{Wigner6j}(j_4,j_5,j_3,j_1,j_2,j_6)
+          &=\operatorname{Wigner6j}(j_1,j_5,j_6,j_4,j_2,j_3)\\
+          &=\operatorname{Wigner6j}(j_4,j_2,j_6,j_1,j_5,j_3)\\
+          &=\operatorname{Wigner6j}(j_4,j_5,j_3,j_1,j_2,j_6)
+         \end{aligned}
```
Comment (jfbu) [inline sympy/physics/wigner.py:549]: In this file the example right  before already used aligned, so I only copied.

File: doc/src/conf.py [hunk 8]
```diff
@@ -372,7 +372,13 @@
...
+''',
+    'sphinxsetup': 'verbatimforcewraps',
```
Comment (jfbu) [inline doc/src/conf.py:381]: log file continues to contain the few related `Overfull \hbox` (https://github.com/sphinx-doc/sphinx/issues/12722) but the problem is actually solved in PDF.  But the vast majority of disappearing contents is caused by https://github.com/sphinx-doc/sphinx/issues/12723.  There is also the problem of the index, but it is solved by https://github.com/sympy/sympy/pull/26866.

File: doc/src/conf.py [hunk 9]
```diff
@@ -372,7 +372,13 @@
 \usepackage{letltxmacro}
 \LetLtxMacro\OldLaTeX\LaTeX
 \AtBeginDocument{\DeclareRobustCommand{\LaTeX}{\text{\OldLaTeX}}}
-'''
+\let\OldUnderscore\_
```
Comment (jfbu) [inline doc/src/conf.py:375]: `\_` in LaTeX is a "robust" macro.  For replacing it I will use the more modern `\protected\def` mechanism.

File: doc/src/conf.py [hunk 10]
```diff
@@ -372,7 +372,13 @@
...
+\makeatletter
+\AtBeginDocument{\sbox\sphinxcontinuationbox{\spx@opt@verbatimcontinued}}
```
Comment (jfbu) [inline doc/src/conf.py:377]: this `\AtBeginDocument` is probably not needed but I think LaTeX does font related things there, so I am delaying to that location.

**Comment by jfbu** [comment]
A few more examples of Attributes tables. 

Capture d’écran 2024-08-01 à 20 23 39

and

Capture d’écran 2024-08-01 à 20 24 04

and

Capture d’écran 2024-08-01 à 20 25 03

It is possible for the last two that the current tabulary rendering is ok, as the descriptions are not too long.  But as soon as descriptions become lengthy, the attribute column will be set by tabulary at the minimal width allowed, which is 40pt if I recall correctly in Sphinx configuration, no it is 3 characters:
```latex
% configuration of tabulary
\setlength{\tymin}{3\fontcharwd\font`0 }% minimal width of "squeezed" columns
``` 

Another approach could be to set `\tymin` to some higher value (say 2cm) but it is needed to allow the linebreaks at underscore as I did. 

No wait, actually I had already decied `tabulary` was no good.  I observed many examples where it was shifted to next page, leaving a big blank because there was not enough left room.  No, `longtable` is better.

One last remark is that the underscore rendering by LaTeX is poor, it is too thin.  This could be improved by yet more latex...

**Comment by jfbu** [comment]
Good surprise: now that wrapping is allowed at underscores some signatures are less catastrophic:
Capture d’écran 2024-08-01 à 20 38 32

Much better but not perfect of course yet. Clearly we must shift all of that to underneath the object name to have more horizontal room.

**Comment by asmeurer** [comment]
Adding a flag to latex() to do line breaks would be independently useful. I don't know how easy it would be, though.

**Comment by asmeurer** [comment]
There's conflicts here now after the other PR was merged, but is this otherwise ready for review?

**Comment by jfbu** [comment]
> There's conflicts here now after the other PR was merged, but is this otherwise ready for review?

I have merged master.  Yes, ready for review.  I may have an idea of a pure LaTeX hack to automatically allow line breaks at symbols such as `=`, `<` and `>`, but this requires testing (I know it would work in pure TeX `$...$` but I need to check in `equation`; can't do until some time, and is only partially related to this PR).

**Comment by jfbu** [comment]
@moorepants @asmeurer The idea I had in mind does not work in `equation/equation*` environments.  And it is not needed in simple `$...$` inline math (or rather `\(...\)` as official in LaTeX, and used for rendering the math *role*). In other terms if you use Sphinx math *role*, not *directive*, then TeX will automatically break equations as symbols such as `=`, `<`, `>`.  TeX leaves them at end of line it seems always.

Capture d’écran 2024-08-02 à 20 36 00

As long as equations are not numbered, using this trick of rendering them using the math *role* may be a simple method to try.

If we really want `equation` environment with or without numbering, following syntax can be used in LaTeX:

```latex
\begin{equation}\vcenter{\(zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz\)}
\end{equation} 
```
 which produces:
Capture d’écran 2024-08-02 à 20 46 38

where equation number is shifted below because displayed occupied all width. Hmm, here I go hacking LaTeX again, you can do that

```latex
\begin{equation}\vcenter{\hsize\dimexpr\linewidth-1cm\relax\(zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz\)}
\end{equation} 
```

which gives
Capture d’écran 2024-08-02 à 20 49 05

But Sphinx by default adds split environment, so I tried

```latex
\begin{equation}\begin{split}\vcenter{\hsize\dimexpr\linewidth-1cm\relax\(zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz>zzzzz\)}
\end{split}\end{equation} 
```

which gives
Capture d’écran 2024-08-02 à 20 50 18

For some reason interline spacing got increased.

In summary a possible thing to try out would be to enclose all math material destined to the math *directive* intro the following structure:
- before: `\vcenter{\hsize\dimexpr\linewidth-1cm\relax\(\displaystyle`
- after math contents: `\)}`

This change looks simple to implement but I am not in the best position to try it out myself.

No idea how matrices environments of amsmath render in such things.  The idea requires testing on real-life examples.

It is a hacky thing. If it does not work I don't think I would try to make it work with additional hackier things (as the amsmath math environments are nasty beasts, which moreover may receive upstream LaTeX evolutions due to their tagging project, I will certainly not myself try to do anything a bit hacky with them as life is finite).

**Comment by asmeurer** [comment]
I expect the problem Jason is talking about will need to be fixed in the way jupyter-sphinx renders math output from SymPy. Maybe we should move this discussion to an issue on that repo. Note that SymPy currently doesn't even use jupyter-sphinx (I think Jason is having this issue with other projects).

**Comment by moorepants** [comment]
Yes, Aaron is correct. I should have opened an issue on jupyter-sphinx and tagged you in there instead of mixing into this thread but at first I mistakenly thought it was a similar issue to wrapping the many digit pi.
